CSharpMath.Editor/Keyboards/MathKeyboard.cs

```diff
@@ -97,20 +97,45 @@ public MathListIndex InsertionIndex
         var subIndexType = isSuperScript ? MathListSubIndexType.Superscript : MathListSubIndexType.Subscript;
         IMathList GetScript(IMathAtom atom) => isSuperScript ? atom.Superscript : atom.Subscript;
         void SetScript(IMathAtom atom, IMathList value) { if (isSuperScript) atom.Superscript = value; else atom.Subscript = value; }
-        if (_insertionIndex.AtBeginningOfLine) {
+        void CreateEmptyAtom() {
           // Create an empty atom and move the insertion index up.
           var emptyAtom = MathAtoms.Placeholder;
           SetScript(emptyAtom, MathAtoms.PlaceholderList);
           MathList.InsertAndAdvance(ref _insertionIndex, emptyAtom, subIndexType);
+        }
+        static bool IsFullPlaceholderRequired(IMathAtom mathAtom) {
+          var atomType = mathAtom.AtomType;
+          switch (atomType) {
+            case MathAtomType.Variable:
+            case MathAtomType.Number:
+            case MathAtomType.Radical:
+            case MathAtomType.Fraction:
+            case MathAtomType.Ordinary:
+            // functions are treated as a LargeOperator so no special case here. Yet?
+            case MathAtomType.LargeOperator:
```

Comment:
`LargeOperators` have special handling for superscripts and subscripts. They always support them.